src/System.Linq/tests/SingleTests.cs

```diff
@@ -0,0 +1,46 @@
+﻿using System.Linq.Tests.Helpers;
+using Xunit;
+
+namespace System.Linq.Tests
+{
+    public class SingleTests
+    {
+        [Fact]
+        public void FailOnEmpty()
+        {
+            Assert.Throws<InvalidOperationException>(() => Enumerable.Empty<int>().Single());
+        }
+
+        [Fact]
+        public void DefaultOnEmpty()
+        {
+            Assert.Equal(0, Enumerable.Empty<int>().SingleOrDefault());
+        }
+
+        [Fact(Skip="Runs for thousands of years. Then throws wrong exception")]
+        public void FailOnManyMatches()
+        {
+            Assert.Throws<InvalidOperationException>(() => Enumerable.Range(0, int.MaxValue).Single(i => i > 0));
```

Comment:
The nice thing about adding them now is that it locks in the behavior and alerts us to it ever changing, on purpose or accidentally.  Kind of like the Assert.Same/NotSame tests you suggested adding around ToArray.  Then any PR that did want to change the short-circuiting behavior would also need to update the test; lack of doing so would make the contributor aware of something changing in case they weren't already, and the fact that the changed tests would then need to be in a passing PR would make the reviewer(s) aware of the potential break as well.